### PR TITLE
Trpc changes

### LINKED ISSUE #77: Investigate Upgrading to Next.js 13
# Overview

This issue is tracking work to investigate potentially upgrading Jupiter to use Next.js 13. This major version update could allow significant benefits, though we need to investigate if it has basic compatibility with the existing architecture, and a potential upgrade path to take advantage of new features. New features of particular interest are server-side components and Turbopack.

For potential architectural/data changes, consideration should be given to the long-term plan for the API team to manage Jupiter's backend, so coordination would be needed to ensure they can support what our frontend is relying on.

# Proposal

- [ ] Investigate new features and potential application to Jupiter
- [ ] Ensure basic compatibility with the existing implementation
- [ ] Incrementally update file structure
- [ ] Incrementally update data/backend structure

## Supplementary Information

Blog post with more info: https://nextjs.org/blog/next-13

### LINKED ISSUE #104: Enable the search bar in the Directory
# Overview

As a user, I should be able to search for orgs using a search field in the Directory page. This is a critical feature for a basic use case.

Likely, the search functionality should consider the title of each org, and potentially the description. For performance reasons, we should consider what data we want to search.

For this issue, typing into the search field should reduce the number of results shown in the Directory. The search could happen after the user presses the enter key, or after the user has stopped typing for x seconds (likely both should be supported). Autocomplete is not needed at this stage. Additionally, animating the results changing would be nice to have but is not critical for this initial implementation.
# Proposal

- [ ] Implement basic backend support
- [ ] Enable filtering displayed orgs
- [ ] Filter based upon text entered into search bar

### LINKED ISSUE #106: User account support needed
# Overview

As a user or an administrator, I should be able to access Jupiter with varying permissions enabled.

Once we have basic backend integration working, it'll be necessary to establish a structure and functionality for user accounts. This way, we can support saving data and allow certain users permission to edit org data.

# Proposal

- [ ] Establish account roles (ex: User, Org owner, Org rep, admin, etc)
- [ ] Implement user roles in the backend
- [ ] Wire up UI visibility and access according to user role

### PR DESCRIPTION
**Pull request by RubenOlano**
Initial changes to try and use trpc for our API

### CODE REVIEW

**Comment by Jake3231** [comment]
Thanks for making this change! Please request review from Jupiter once it’s good to go, to be sure it can get merged.

**Comment by Buchimi** [comment]
Also, this is quite a loaded pull request. It'll take me a while to comb through all of it.
There also seems to be conflicts with the test file.
Also, it looks like the deployment failed, you should take a look into that

**Comment by RubenOlano** [comment]
> trpc? what was your reasoning behind this?

For now, the reason why I considered using TRPC was for connecting the frontend to the backend and using it as a wrapper over React Query for any fetch requests we make. The current implementation of the search algorithm makes a request every time the user types something ("N" and "Neb" do two different requests) depending on typing speed. This helps just cache any requests made so we're not overloading the browser with requests

**Comment by RubenOlano** [comment]
> Also, this is quite a loaded pull request. It'll take me a while to comb through all of it. There also seems to be conflicts with the test file. Also, it looks like the deployment failed, you should take a look into that

Yeah, sorry about that I just forgot to mark it for review a while ago. As for the failing deployment, that is mainly just on the new environment variables to use with next-auth. I'll look into the test confclit

**Comment by Buchimi** [comment]
Oh you were able to integrate next auth? I thought that pull request was cancelled.

**Comment by RubenOlano** [comment]
> Oh you were able to integrate next auth? I thought that pull request was cancelled.

I canceled that pr just because this pr also implements it. The benefit of implementing it here is that we can check that any trpc requests that are made can have valid auth validation.

**Comment by Jake3231** [comment]
Re this change, we'll likely need to find a way to move the keys from the json cert file into the `env` in order to deploy to Vercel correctly. It looks like we should be able to change the way we're loading the Google auth keys without much other change

**Comment by RubenOlano** [comment]
Shouldn't be too much of a change, I'll just update the env schema and we should be able to work, just need to make sure that if these are ever leaked that we're able to change them on vercel

File: .eslintrc.cjs
Comment (Jake3231) [inline .eslintrc.cjs:1]: Is this adding linting to the project?

File: src/backend_tools/db_provider.ts
Comment (Jake3231) [inline src/backend_tools/db_provider.ts:27]: What's the reason to remove the type annotation?

File: src/backend_tools/db_provider.ts
Comment (Jake3231) [inline src/backend_tools/db_provider.ts:85]: Is there potentially a way to query by name from the database, rather than by fetching all clubs and filtering them locally?

File: src/backend_tools/db_provider.ts
Comment (Jake3231) [inline src/backend_tools/db_provider.ts:126]: Is ID just name?

File: src/components/navigation/Navbar.tsx
Comment (Jake3231) [inline src/components/navigation/Navbar.tsx:3]: What is the reason to switch to jpg here?

File: src/env.mjs
Comment (Jake3231) [inline src/env.mjs:78]: Is this requiring the cert file to be a large JSON string inside the env?

File: src/pages/api/trpc/[trpc].ts
Comment (Jake3231) [inline src/pages/api/trpc/[trpc].ts:12]: Will this still work in production?

File: src/server/api/routers/example.ts
Comment (Jake3231) [inline src/server/api/routers/example.ts:1]: Do we need to keep this file?

File: .eslintrc.cjs
Comment (RubenOlano) [reply .eslintrc.cjs:1]: I added this, but we can update based on engineering requirements once those are out

File: src/backend_tools/db_provider.ts
Comment (RubenOlano) [reply src/backend_tools/db_provider.ts:85]: Unfortunatly you can't do that with firestore

File: src/backend_tools/db_provider.ts
Comment (RubenOlano) [reply src/backend_tools/db_provider.ts:126]: That's just a remnant from the original implemetation, I'll change that whenever we are able to implement actual events

File: src/components/navigation/Navbar.tsx
Comment (RubenOlano) [reply src/components/navigation/Navbar.tsx:3]: no particular reason, I just changed it on my local branch

File: src/env.mjs
Comment (RubenOlano) [reply src/env.mjs:78]: The env will appear as normal, but the processEnv object is what's going to be used whenver trying to access any environment variables on the server

File: src/pages/api/trpc/[trpc].ts
Comment (RubenOlano) [reply src/pages/api/trpc/[trpc].ts:12]: This just means that we get extra logs whenever running on dev server. On live server the logs won't appear as verbose

File: src/server/api/routers/example.ts
Comment (RubenOlano) [reply src/server/api/routers/example.ts:1]: Probably good to keeps just as an example for other devs, but we can remove it

File: src/backend_tools/db_provider.ts
Comment (RubenOlano) [reply src/backend_tools/db_provider.ts:27]: It automatically type annotates it under the hood so it's just not necessary

File: src/server/api/routers/example.ts
Comment (Jake3231) [reply src/server/api/routers/example.ts:1]: Sounds good, we'll just want to be sure to mention that this adds linting also

**Comment by Jake3231** [comment]
Also adds linting to the repo.

File: package.json
Comment (Jake3231) [inline package.json:27]: I missed that this upgrades to Next 13

**Review by Jake3231 — DISMISSED** [review]
Overall looks great, thanks for making this change! For this branch, I see some changes, though does it still enable search on the homepage on the new design?

File: src/env.mjs
Comment (RubenOlano) [reply src/env.mjs:78]: Does this make sense?

Comment (Jake3231) [reply src/env.mjs:78]: Yeah, that makes sense

**Review by Shaurya0108 — DISMISSED** [review]
TRPC makes the backend communication a lot better!

File: package.json
Comment (jasonappah) [inline package.json:11]: Do we need to keep these around? My understanding is that this gets done automatically with Next.js' dev server

Comment (RubenOlano) [reply package.json:11]: If that's the case then I can just remove that in a future pr

File: package.json
Comment (jasonappah) [inline package.json:25]: Do we need to keep this dep around?

File: package.json
Comment (jasonappah) [inline package.json:21]: Could we move this + `mocha` to be dev dependencies?

File: package.json
Comment (jasonappah) [inline package.json:29]: If we don't use those PostCSS scripts we can remove this dep

**Review by jasonappah — DISMISSED** [review]
These changes look great - just had a few questions here and there but nothing that should block merging if you want to do that.

One other thing that I didn't mention in another comment: can we also remove `/workspaces/jupiter/src/styles/Home.module.css` and `/workspaces/jupiter/src/styles/main.css`? I couldn't see any other references to these files but I could be missing something

File: package.json
Comment (jasonappah) [reply package.json:25]: oops meant to highlight the `isomorphic-unfetch` dep here

File: package.json
Comment (jasonappah) [reply package.json:21]: this was supposed to be highlighting `chai`

File: package.json
Comment (jasonappah) [reply package.json:29]: this was also supposed to be highlighting `postcss-cli`

File: package.json
Comment (RubenOlano) [reply package.json:25]: Fuse is going to be used to fuzzy search for orgs

**Review by jasonappah — APPROVED** [review]

**Pull request merged**
